pkg/instancetype/controller/vm/controller_test.go

```diff
@@ -0,0 +1,1052 @@
+		})
+
+		It("should store VirtualMachinePreference as ControllerRevision on Sync", func() {
+			vm.Spec.Preference = &virtv1.PreferenceMatcher{
+				Name: preference.Name,
+				Kind: instancetypeapi.SingularPreferenceResourceName,
+			}
+
+			var err error
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+			Expect(err).To(Succeed())
+
+			expectedPreferenceRevisionName := instancetype.GetRevisionName(
+				vm.Name, preference.Name, preference.GroupVersionKind().Version, preference.UID, preference.Generation)
+			expectedPreferenceRevision, err := instancetype.CreateControllerRevision(vm, preference)
+			Expect(err).ToNot(HaveOccurred())
+
+			sanitySync(vm, vmi)
+
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+			Expect(err).To(Succeed())
+			Expect(vm.Spec.Preference.RevisionName).To(Equal(expectedPreferenceRevision.Name))
+
+			preferenceRevision, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(
+				context.Background(), expectedPreferenceRevisionName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			preferenceRevisionObj, ok := preferenceRevision.Data.Object.(*v1beta1.VirtualMachinePreference)
+			Expect(ok).To(BeTrue(), "Expected Preference in ControllerRevision")
+			Expect(preferenceRevisionObj.Spec).To(Equal(preference.Spec))
+		})
+
+		DescribeTable("should apply VirtualMachinePreference from ControllerRevision to VirtualMachineInstance",
+			func(getRevisionData func() []byte) {
+				preferenceRevision := &appsv1.ControllerRevision{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "crName",
+					},
+					Data: runtime.RawExtension{
+						Raw: getRevisionData(),
+					},
+				}
+
+				preferenceRevision, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(
+					context.Background(), preferenceRevision, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				vm.Spec.Preference = &virtv1.PreferenceMatcher{
+					Name:         preference.Name,
+					Kind:         instancetypeapi.SingularPreferenceResourceName,
+					RevisionName: preferenceRevision.Name,
+				}
+
+				vm, err = virtClient.VirtualMachine(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+				Expect(err).To(Succeed())
+
+				Expect(instancetypeController.ApplyToVMI(vm, vmi)).To(Succeed())
+
+				Expect(vmi.Spec.Domain.Firmware.Bootloader.EFI).ToNot(BeNil())
+				Expect(vmi.Annotations).ToNot(HaveKey(virtv1.InstancetypeAnnotation))
+				Expect(vmi.Annotations).To(HaveKeyWithValue(virtv1.PreferenceAnnotation, preference.Name))
+				Expect(vmi.Annotations).ToNot(HaveKey(virtv1.ClusterInstancetypeAnnotation))
+				Expect(vmi.Annotations).ToNot(HaveKey(virtv1.ClusterPreferenceAnnotation))
+			},
+			Entry("using v1alpha1 and VirtualMachinePreferenceSpecRevision with APIVersion", func() []byte {
+				v1alpha1preferenceSpec := v1alpha1.VirtualMachinePreferenceSpec{
```

Comment:
Would be possible to extract the v1alpha1.VirtualMachinePreferenceSpec struct to somewhere and then just reuse it in the other 2 tests?

Reply:
As above.